File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import heapq
import sys
input = sys.stdin.readline

INF = int(1e9)
n = int(input().rstrip())
m = int(input().rstrip())
gragh = [[] for _ in range(n+1)]
distance = [INF] * (n+1)

for _ in range(m):
    u, v, w = map(int, input().rstrip().split())
    gragh[u].append((v, w))
    
start, end = map(int, input().rstrip().split())

def dijkstra(start):
    pq = []
    heapq.heappush(pq, (0, start))
    distance[start] = 0
    
    while pq:
        dist, node = heapq.heappop(pq)
        
        if distance[node] < dist:
            continue
        
        for i in gragh[node]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(pq, (cost, i[0]))
                
dijkstra(start)
                
print(distance[end])
    

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.java
import java.io.*;
import java.util.*;

class Node implements Comparable<Node> {
	int idx;
	int cost;

	Node(int idx, int cost) {
		this.idx = idx;
		this.cost = cost;
	}

	@Override
	public int compareTo(Node another) {
		return this.cost - another.cost;
	}
}

public class Main {
	static List<List<Node>> graph = new ArrayList<>();
	static int[] distance;

    public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int V = Integer.parseInt(br.readLine());
		int E = Integer.parseInt(br.readLine());
		distance = new int[V + 1];
		Arrays.fill(distance, Integer.MAX_VALUE);

		for (int i = 0; i < V + 1; i++) {
			graph.add(new ArrayList<>());
		}

		for (int i = 0; i < E; i++) {
			String[] nodeInfo = br.readLine().split(" ");
			int a = Integer.parseInt(nodeInfo[0]);
			int b = Integer.parseInt(nodeInfo[1]);
			int c = Integer.parseInt(nodeInfo[2]);
			graph.get(a).add(new Node(b, c));
		}

		String[] info = br.readLine().split(" ");
		int departure = Integer.parseInt(info[0]);
		int arrival = Integer.parseInt(info[1]);

		dijkstra(departure);
		System.out.println(distance[arrival]);
	}

	static void dijkstra(int departure) {
		Queue<Node> pq = new PriorityQueue<>();
		distance[departure] = 0;
		pq.offer(new Node(departure, 0));

		while (!pq.isEmpty()) {
			Node cur = pq.poll();

			if (distance[cur.idx] < cur.cost) continue;

			for (Node next : graph.get(cur.idx)) {
				if (distance[next.idx] > distance[cur.idx] + next.cost) {
					distance[next.idx] = distance[cur.idx] + next.cost;
					pq.offer(new Node(next.idx, distance[next.idx]));
				}
			}
		}
	}
}